pkg/action/install.go

```diff
@@ -368,63 +371,76 @@ func (i *Install) RunWithContext(ctx context.Context, chrt *chart.Chart, vals ma
 		// not working.
 		return rel, err
 	}
-	rChan := make(chan resultMessage)
-	ctxChan := make(chan resultMessage)
-	doneChan := make(chan struct{})
-	defer close(doneChan)
-	go i.performInstall(rChan, rel, toBeAdopted, resources)
-	go i.handleContext(ctx, ctxChan, doneChan, rel)
-	select {
-	case result := <-rChan:
-		return result.r, result.e
-	case result := <-ctxChan:
-		return result.r, result.e
+
+	rel, err = i.performInstallCtx(ctx, rel, toBeAdopted, resources)
+	if err != nil {
+		rel, err = i.failRelease(rel, err)
 	}
+	return rel, err
 }
 
-func (i *Install) performInstall(c chan<- resultMessage, rel *release.Release, toBeAdopted kube.ResourceList, resources kube.ResourceList) {
+func (i *Install) performInstallCtx(ctx context.Context, rel *release.Release, toBeAdopted kube.ResourceList, resources kube.ResourceList) (*release.Release, error) {
+	type Msg struct {
+		r *release.Release
+		e error
+	}
+	resultChan := make(chan Msg, 2)
+	doneChan := make(chan struct{})
+	defer close(doneChan)
+
+	go func() {
+		rel, err := i.performInstall(rel, toBeAdopted, resources)
+		resultChan <- Msg{rel, err}
+	}()
+	go func() {
+		select {
+		case <-ctx.Done():
+			err := ctx.Err()
+			resultChan <- Msg{rel, err}
+		case <-doneChan:
+			return
+		}
+	}()
```

Comment:
When the context is canceled the second goroutine will exit but it looks like the first will still run until `performInstall` reaches completion. Am I missing something?

I wonder if `performInstall` needs to be modified to accept a context and regularly check if it's done to exit.

Reply:
> When the context is canceled the second goroutine will exit but it looks like the first will still run until performInstall reaches completion.

This is right - it was the original behavior even before this PR. This was captured by the added comment to the function. I think it's a technical debt.

> I wonder if performInstall needs to be modified to accept a context and regularly check if it's done to exit.

Yes. The API communicating with K8s (like `KubeClient.WaitWithJobs`) would need to take context instead of timeout, some people consider this in https://github.com/helm/helm/issues/12086

I think that this PR might be a good starting point to fix the mentioned issue (separetes the code into `releasingUpgradeCtx` and `performInstallCtx`).

Reply:
I think this PR (as is) fixes lost goroutines in performInstall/Upgrade.

https://github.com/helm/helm/issues/12086 (and associate PR: https://github.com/helm/helm/pull/12109) might be a better place to tackle synchronizing install.upgrades to not continue to run in the background?

Reply:
I think this PR (as is) fixes lost goroutines in performInstall/Upgrade.

https://github.com/helm/helm/issues/12086 (and associate PR: https://github.com/helm/helm/pull/12109) might be a better place to tackle synchronizing install.upgrades to not continue to run in the background?